### ISSUE TITLE
Refactoring the dataflow of CreateFspec

### ISSUE THREAD

**Issue by stefjoosten** [AmpersandTarski/Ampersand#971]
@hanjoosten and @stefjoosten ran into the need to refactor the function `createMulti :: Options -> IO(Guarded MultiFSpecs)` from module `CreateFspec`. Neither of us could solve issue #967 because this function is too complex to analyse.

#### Version of ampersand that was used 
Ampersand-v3.17.0 [data-analysis:e4d946007*]

#### Problem
It was too complicated to analyse the code of the mentioned function. Here is part of the analysis:
![IMG_20190528_183541](https://user-images.githubusercontent.com/275768/58495492-f26f9100-8177-11e9-90a9-0315ebfdb4ad.jpg). I called Han for explanation

#### Analysis
We decided to rethink the workflow, because the different command-line options and the various artifacts justify thinking before doing. We also decided we don't have the answers yet, so opening a dialogue seems like a good idea. Here is my first shot at the internal data flow of the compiler (Han, your turn):

![Compiler Architecture](https://user-images.githubusercontent.com/275768/58495747-85103000-8178-11e9-8a45-9c9547b7d891.png)

**Comment by stefjoosten** [comment]
So what does this do for the command-line options?

* The option `--meta-tables` becomes obsolete.
* The option `--meta-file` has overlap with `--export`.
* The option `--add-semantic-metamodel` becomes obsolete.
* A new option `--metamodel` can be used to generate an Atlas.
* The option `--gen-as-rap-model` need reconsideration.
* The option `--atlas-without-expressions` need reconsideration.

**Comment by stefjoosten** [comment]
### Analysis
There is an interesting pattern here. Let us look for the similarities in three streams: SystemContext, UserScript, and Documentation. Here is a first take at it:

![AmpersandCompilerArchitectur-Page-1](https://user-images.githubusercontent.com/275768/58744703-38fa0f80-8446-11e9-83fa-93f7e6920377.png)

The three are nearly similar. Look at the metamodel, `AST.adl`, in the middle lane. We use AST as a mold, which we fill with the contents of the user script.  That yields (in the end) an atlas. Making a prototype works similarly. The mold is the user script itself together with `SystemContext.adl`. We fill it with the contents of the user script. We can apply the exact same idea to generate the documentation. We can use `Documentation.adl` (yet to be made) as mold and fill it with the contents of the user script.

One more remark on prototype generation. The mold is the user script together with SystemContext. That is different from the other two streams, where the mold is entirely specified by `AST.adl` and `Documentation.adl` respectively.

**Comment by stefjoosten** [comment]
Analysis 2: Making the "shape in a mold" process into a function

![AmpersandCompilerArchitectur-Page-2 (1)](https://user-images.githubusercontent.com/275768/58745099-a3ae4980-844c-11e9-92a9-ba98bf2d22cc.png)

comment to be added

**Comment by hanjoosten** [comment]
The above thinking shows the need for composable functions. Instead of drawing them, lets do this with types.

First of all, we know that we have different kind of metamodels, and this could be more in the future (@stefjoosten has idead about FADocumented)

~~~haskell
data MetaModel = FormalAmpersand | SystemContext
~~~

Each metamodel comes with its .adl script. It also comes with a function in Haskell that tells how to populate each relation in that metamodel with content from a (users)script. This knowledge can be stored in a datastructure called `grindInfo`. The ampersand-compiler buids itself a `Map MetaModel GrindInfo`. 

To do all grinding, we now have the following composable building blocks:

~~~haskell
-- | The 'grind' function creates a P_Context that contains the population for every relation in the metamodel. The population is defined by the given FSpec, which usually is the FSpec of the user. 
grind :: Options -> GrindInfo -> FSpec -> P_Context

-- | The 'addsemanticModel' basically does a union of the P_Context of the .adl script of the metamodel with the P_Context of the user's script. It should have the same effect as if the user had specified an INCLUDE statement referencing to the .adl file of the metamodel.
addSemanticModel :: GrindInfo -> P_Context -> P_Context
~~~

That's it. Using these basic functions, we can build whatever FSpec we need. Have a look in the code for more details.

**Comment by stefjoosten** [comment]
Good idea to do this with types. However, your comment does not show how to do "this" (i.e. Analysis 2, above) with types. Could you please explain rather than letting me puzzle over code?

I am curious to read how `createMulti` facilitates an arbitrary number of streams generically.

**Comment by stefjoosten** [comment]
For one thing, I noticed that `--dataAnalysis` works no longer. I don't know whether this has something to do with `createMulti`, but it is definitely one of the streams that we want to be able to do.

**Comment by Michiel-s** [comment]
Thanks both of you for clearing this up and improving the code. It is/was kinda messy.

The distinction @hanjoosten makes is very good I think, because we need combinations for different purposes:
* RAP3 application: addSemanticModel FormalAmpersand
* Population for RAP: grind + addSemanticModel with FormalAmpersand
* All prototypes: grind + addSemanticModel with SystemContext

**Comment by Michiel-s** [comment]
@hanjoosten refactored the whole meat grinder and build recipes some time ago. That's working good now and the code is less complex and more readable. Well done!

I think we only need to document this. The diagrams above are very useful for this.

So let's document and then close this issue here. I'll help with that.
